Spack V1 Migration Feature:  Reserved Spack Definitions

CodeGat opened the pull request:
Closes #330 
Will be merged into #326 

## Background

Root specs are very meaningful in our infrastructure. This comes with a few downsides, mostly in the cases where we have multiple specs in the speclist, or regular users want to use prerelease manifests in their own instances but can't due to prerelease-specific modifications of the root spec. 

In this PR, we move away from important information being in the `.spack.specs[0]` field, and instead use reserved, dummy definitions at the top of the file:

```yaml
spack:
  definitions:
  - _name: [access-om2]
  - _version: [2025.11.000]
  specs:
  - access-om2
  # ...
```

This means that speclists can be free from infra-specific information, and they are open to future modification.

## The PR

* Rename the `get-spack-root-spec` action to the more apt `get-spack-manifest`
  * Also update the outputs - simplified to `deployment-{name|version}` outputs, equivalent to `.spack.definitions[]._{name|version}[0]` fields. 
* Update all usages of the `get-spack-root-spec` action to `get-spack-manifest` 
* Update the internals of injection to allow multiple of the same spec, using this rule:
  * If there are no projections defined, make them of the form `{name}/prX-Y/{hash:7}`
  * If there are projections defined, infix the deployment number. This is of the form: `{name}/prX-Y/DEMARCATOR`, where the DEMARCATOR is the defined projection. For example, `{name}/prX-Y/{variants.x}x{variants.y}`
  * This is the same for Releases, except using the `_version`. It is not yet allowed in Releases.  

## Testing

<details>
<summary>Testing Details in this collapsed section</summary>

### Unit Tests

Done via `pytest`, all pass. 

### MDR Testing

Done for regular MDRs in ACCESS-NRI/ACCESS-TEST#60

#### Prerelease (Draft)

Success, see https://github.com/ACCESS-NRI/ACCESS-TEST/actions/runs/19916256065

#### Prerelease (Ready For Review)

Success, see https://github.com/ACCESS-NRI/ACCESS-TEST/actions/runs/20013811438

#### Prerelease (`!redeploy`)

Expected failure, since the schema version updates are not yet on the `main` branch (since comment triggers are technically in the context of the `main` branch). But, the only points of difference (The status setting in https://github.com/ACCESS-NRI/ACCESS-TEST/actions/runs/20014778511/job/57390074093 and https://github.com/ACCESS-NRI/ACCESS-TEST/actions/runs/20014778511/job/57390091047) succeeded as expected. 

### Software Deployment Repository Testing

Done for Software Deployment Repostiories in ACCESS-NRI/system-tools#19

#### Prerelease (Draft)

Succeeded, see https://github.com/ACCESS-NRI/system-tools/actions/runs/20151461268?pr=19

#### Prerelease (Ready For Review)

Succeeded. see https://github.com/ACCESS-NRI/system-tools/actions/runs/20217336546?pr=19

#### Prerelease (`!redeploy`)

Expected failure, since the schema version updates are not yet on the `main` branch (since comment triggers are technically in the context of the `main` branch). But, the only points of difference (The status setting in https://github.com/ACCESS-NRI/system-tools/actions/runs/20217823516/job/58033899157 and https://github.com/ACCESS-NRI/system-tools/actions/runs/20217823516/job/58033912451) succeeded as expected. 

</details>

CodeGat committed:
Rename action from get-spack-root-spec to get-spack-manifest, return only deployment-[name|version]

CodeGat committed:
cd.yml: Use get-spack-manifest, rework job names

CodeGat committed:
ci-comment.yml: Use get-spack-manifest, simplify update and commit of the bump

CodeGat committed:
deploy-1-setup.yml: Use get-spack-manifest, remove check on root spec usage

CodeGat committed:
deploy-2-start.yml: Remove unused manifest information

CodeGat committed:
fix: Add manifest path to action

CodeGat committed:
Update modules.py, prerelease.py to use reserved defs _name and _version

CodeGat committed:
Update tests

CodeGat committed:
Make reserved definitions flow-style for module injection too

CodeGat committed:
Remove references to RootSpecs, use ReservedDefinitions.get("name"), remove --keep-root-spec-intact flag

CodeGat commented:
This looks about ready to go!

CodeGat committed:
Update pkg_repo_url logic to use spack.repo.PATH.get_pkg_class()

aidanheerdegen commented on scripts/spack_manifest/getter.py:
```diff
@@ -16,14 +17,78 @@ class NoSectionComponentError(Exception):
+        #   {'something': ['else']}
+        # ]}
+        # Into a much easier to parse:
+        # ['name': 'access-om2', 'version': '2025.02.100']
```
```suggestion
        # [{'name': 'access-om2'}, {'version': '2025.02.100'}]
```

aidanheerdegen replied:
Speaking of which, why an array of single key/value dicts, and not just a dict with multiple key/value pairs?

CodeGat replied:
Whoops, they actually were `dict[str, Any]` anyway! Done in https://github.com/ACCESS-NRI/build-cd/pull/332/commits/b60c71d992d604ada33664cd19ec619982282a30

aidanheerdegen commented on scripts/spack_manifest/getter.py:
```diff
@@ -16,14 +17,78 @@ class NoSectionComponentError(Exception):
+                            reserved_definitions[reserved_name_no_underscore] = (
+                                reserved_value_list[0]
+                            )
+
```
If the other case isn't being handled currently why bother with this logic? Could just mention in a comment.

CodeGat replied:
Fair enough, done in https://github.com/ACCESS-NRI/build-cd/pull/332/commits/b60c71d992d604ada33664cd19ec619982282a30

aidanheerdegen commented on scripts/spack_manifest/getter.py:
```diff
@@ -170,7 +235,9 @@ def get_all_package_names_with_ref_requirement(self) -> list[str]:
+            if len(package_spec.get("require", {})) != 0 and package_spec["require"][
+                0
+            ].startswith("@"):
                 packages_with_versions_defined.append(package_name)
```
Don't know if this was an automatice code-formatter, but that is horrendous line breaking.

CodeGat replied:
Thanks, `black`. Done in https://github.com/ACCESS-NRI/build-cd/pull/332/commits/b60c71d992d604ada33664cd19ec619982282a30

aidanheerdegen commented on tests/scripts/spack_manifest/injection/inputs/prerelease.spack.yaml:
```diff
@@ -4,8 +4,11 @@
 spack:
   # add package specs to the `specs` list
+  definitions:
+  - _name: [access-om2]
+  - _version: [2024.03.0]
   specs:
```
This has different indentation to the `spack.yaml` input. Don't know if this is relevant.

CodeGat replied:
It wouldn't affect the tests, but it's good to be consistent - done in https://github.com/ACCESS-NRI/build-cd/pull/332/commits/984b2d3b45bc14047a083628bd09e5a0816907f3

aidanheerdegen commented on tests/scripts/spack_manifest/injection/inputs/spack.yaml:
```diff
@@ -4,8 +4,11 @@
   # add package specs to the `specs` list
+  definitions:
+    - _name: ["access-om2"]
+    - _version: ["2024.03.0"]
```
So we support/test for reserved definitions other than than `_name` and `_version`?

CodeGat replied:
We don't test for them yet, since `_name` and `_version` are the only two I can think of so far.

aidanheerdegen reviewed (commented):
An initial pass, but might need a peer-review to ask some more questions.

CodeGat commented:
After the rebase onto `v7`, I note that it still works fine :+1: See https://github.com/ACCESS-NRI/ACCESS-TEST/pull/60#issuecomment-3672847338

CodeGat committed:
getter.py fixes from code review
Update reserved defs getter return value, remove case statement, revert autoformatter

CodeGat committed:
Update test input manifest formatting

CodeGat committed:
Add back manifest info step

CodeGat committed:
Update pre/release injections to handle multiple specs, simplify custom release projection logic

CodeGat committed:
Change release injection --packages from space-sep to comma-sep, as convention

CodeGat committed:
Update tests

CodeGat commented:
After the rebase, I note it has passed testing: https://github.com/ACCESS-NRI/ACCESS-TEST/pull/60#issuecomment-3752751361

CodeGat committed:
Explicitly disallow multi-spec Releases

CodeGat commented on .github/actions/get-spack-manifest/action.yml:
```diff
@@ -0,0 +1,60 @@
+    description: |
+      The version of the deployment as specified in the reserved definition _version, in the spack manifest file.
+    value: ${{ steps.defs.outputs.version }}
+runs:
+  using: composite
+  steps:
```
Consider getting this info from the getter.py class instead of yq

CodeGat replied:
Done in https://github.com/ACCESS-NRI/build-cd/pull/332/commits/16fc6c6a626f808861267240ada52e3f0e1feea7 - interesting!

CodeGat commented on .github/workflows/ci-comment.yml:
```diff
@@ -131,22 +131,17 @@ jobs:
-          updated_spec="${{ needs.defaults.outputs.root-sbd }}@git.${{ steps.bump.outputs.after }}${{ env.OPTIONAL_VERSION }}"
-          yq -i "${{ steps.original.outputs.yq-root-spec }} = \"$updated_spec\"" spack.yaml
-          yq -i '${{ env.SPACK_YAML_MODEL_PROJECTION_YQ }} = "{name}/${{ steps.bump.outputs.after }}"' spack.yaml
+          yq -i '.spack.definitions[]._name[0] = "${{ steps.bump.outputs.after }}"' spack.yaml
```
`_name` should be `_version`

CodeGat replied:
Done as part of https://github.com/ACCESS-NRI/build-cd/pull/332/commits/1f0d0a4ddc9e7e022e99ba1a56b287476e117283

aidanheerdegen commented on .github/workflows/deploy-1-setup.yml:
```diff
@@ -298,21 +298,19 @@ jobs:
-              echo "::error::The root spec in the spack manifest (${{ steps.current.outputs.root-spec-name }}) cannot be different from the expected root spec (${{ inputs.expected-root-spec-name }}) in non-Draft PRs."
-              exit 1
+          if [ "$root_specs" -gt 1 ]; then
+            echo "::error::We don't yet support multiple root specs with the same package name. Contact the Model Release Team for information."
```
Note this specifically prohibits multiple specs with the same package name as `deployment-name`

CodeGat commented on scripts/spack_manifest/injection/modules.py:
```diff
@@ -72,10 +108,10 @@ def inject_projections(
-        new_projections.update(
-            generate_projection_for_root_spec_or_raise(manifest, root_spec)
-        )
+    # if root_spec not in defined_projections:
```
Remove comment

CodeGat replied:
Done as part of https://github.com/ACCESS-NRI/build-cd/pull/332/commits/1f0d0a4ddc9e7e022e99ba1a56b287476e117283

aidanheerdegen replied:
`modules.py` isn't modified in that commit. Nonetheless I can't find that line, so I guess all good!

CodeGat commented on scripts/spack_manifest/injection/prerelease.py:
```diff
@@ -19,25 +20,35 @@
+
 
 yaml.add_representer(YamlExplicitQuotedString, yaml_explicit_quoted_string_representer)
+yaml.add_representer(YamlExplicitFlowStyleSequence, yaml_explicit_flow_style_sequence_representer)
```
Import the other one from the modules script (or maybe add it to it's own class?)

CodeGat replied:
Done in https://github.com/ACCESS-NRI/build-cd/pull/332/commits/5a1c7e3bc4a4e08d5f4e352a2ecd9283355aa7e5

CodeGat commented on scripts/spack_manifest/injection/prerelease.py:
```diff
@@ -73,13 +80,31 @@ def inject_prerelease_information(
+                if reserved_definition.startswith("_"):
+                    manifest["spack"]["definitions"][i][reserved_definition] = YamlExplicitFlowStyleSequence(reserved_value_list)
+
+    return manifest
```
Also make this an import

CodeGat replied:
Done in https://github.com/ACCESS-NRI/build-cd/pull/332/commits/5a1c7e3bc4a4e08d5f4e352a2ecd9283355aa7e5

CodeGat commented on scripts/spack_manifest/injection/prerelease.py:
```diff
@@ -110,43 +135,34 @@ def add_namespace_to_other_projection_versions(
-    else:
-        updated_version: str = f"{{name}}/{root_spec_version}"
+    if current_root_projection:
+        new_root_projection = f'{{name}}/{deployment_version}'
```
Check if this bit needs to be here

CodeGat replied:
Done as part of https://github.com/ACCESS-NRI/build-cd/pull/332/commits/1f0d0a4ddc9e7e022e99ba1a56b287476e117283

CodeGat commented on scripts/spack_manifest/getter.py:
```diff
@@ -15,6 +15,107 @@ class NoSectionComponentError(Exception):
+        # The multi-target format is of the form:
+        # spack:
+        #   definitions:
+        #     - ROOT_PACKAGE: [access-om2@git.2025.05]
```
No version in comment

CodeGat replied:
Done as part of https://github.com/ACCESS-NRI/build-cd/pull/332/commits/1f0d0a4ddc9e7e022e99ba1a56b287476e117283

CodeGat committed:
get-spack-manifest: call the getter.py script

CodeGat committed:
Import yaml representers from own class

CodeGat committed:
Small fixes

aidanheerdegen reviewed (approved):
LGTM, thanks! Nice work.
